// youtubei.js/dist/src/parser/classes/misc/Format.js
var _Format_this_response_nsig_cache;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { InnertubeError } from '../../../utils/Utils.js';
class Format {
    constructor(data, this_response_nsig_cache) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        _Format_this_response_nsig_cache.set(this, void 0);
        if (this_response_nsig_cache) {
            __classPrivateFieldSet(this, _Format_this_response_nsig_cache, this_response_nsig_cache, "f");
        }
        this.itag = data.itag;
        this.mime_type = data.mimeType;
        this.is_type_otf = data.type === 'FORMAT_STREAM_TYPE_OTF';
        this.bitrate = data.bitrate;
        this.average_bitrate = data.averageBitrate;
        this.width = data.width;
        this.height = data.height;
        this.projection_type = data.projectionType;
        this.stereo_layout = (_a = data.stereoLayout) === null || _a === void 0 ? void 0 : _a.replace('STEREO_LAYOUT_', '');
        this.init_range = data.initRange ? {
            start: parseInt(data.initRange.start),
            end: parseInt(data.initRange.end)
        } : undefined;
        this.index_range = data.indexRange ? {
            start: parseInt(data.indexRange.start),
            end: parseInt(data.indexRange.end)
        } : undefined;
        this.last_modified = new Date(Math.floor(parseInt(data.lastModified) / 1000));
        this.content_length = parseInt(data.contentLength);
        this.quality = data.quality;
        this.quality_label = data.qualityLabel;
        this.fps = data.fps;
        this.url = data.url;
        this.cipher = data.cipher;
        this.signature_cipher = data.signatureCipher;
        this.audio_quality = data.audioQuality;
        this.approx_duration_ms = parseInt(data.approxDurationMs);
        this.audio_sample_rate = parseInt(data.audioSampleRate);
        this.audio_channels = data.audioChannels;
        this.loudness_db = data.loudnessDb;
        this.spatial_audio_type = (_b = data.spatialAudioType) === null || _b === void 0 ? void 0 : _b.replace('SPATIAL_AUDIO_TYPE_', '');
        this.max_dvr_duration_sec = data.maxDvrDurationSec;
        this.target_duration_dec = data.targetDurationSec;
        this.has_audio = !!data.audioBitrate || !!data.audioQuality;
        this.has_video = !!data.qualityLabel;
        this.has_text = !!data.captionTrack;
        this.color_info = data.colorInfo ? {
            primaries: (_c = data.colorInfo.primaries) === null || _c === void 0 ? void 0 : _c.replace('COLOR_PRIMARIES_', ''),
            transfer_characteristics: (_d = data.colorInfo.transferCharacteristics) === null || _d === void 0 ? void 0 : _d.replace('COLOR_TRANSFER_CHARACTERISTICS_', ''),
            matrix_coefficients: (_e = data.colorInfo.matrixCoefficients) === null || _e === void 0 ? void 0 : _e.replace('COLOR_MATRIX_COEFFICIENTS_', '')
        } : undefined;
        if (Reflect.has(data, 'audioTrack')) {
            this.audio_track = {
                audio_is_default: data.audioTrack.audioIsDefault,
                display_name: data.audioTrack.displayName,
                id: data.audioTrack.id
            };
        }
        if (Reflect.has(data, 'captionTrack')) {
            this.caption_track = {
                display_name: data.captionTrack.displayName,
                vss_id: data.captionTrack.vssId,
                language_code: data.captionTrack.languageCode,
                kind: data.captionTrack.kind,
                id: data.captionTrack.id
            };
        }
        if (this.has_audio || this.has_text) {
            const args = new URLSearchParams(this.cipher || this.signature_cipher);
            const url_components = new URLSearchParams(args.get('url') || this.url);
            const xtags = (_f = url_components.get('xtags')) === null || _f === void 0 ? void 0 : _f.split(':');
            this.language = ((_g = xtags === null || xtags === void 0 ? void 0 : xtags.find((x) => x.startsWith('lang='))) === null || _g === void 0 ? void 0 : _g.split('=')[1]) || null;
            if (this.has_audio) {
                const audio_content = (_h = xtags === null || xtags === void 0 ? void 0 : xtags.find((x) => x.startsWith('acont='))) === null || _h === void 0 ? void 0 : _h.split('=')[1];
                this.is_dubbed = audio_content === 'dubbed';
                this.is_descriptive = audio_content === 'descriptive';
                this.is_original = audio_content === 'original' || (!this.is_dubbed && !this.is_descriptive);
            }
            // Some text tracks don't have xtags while others do
            if (this.has_text && !this.language && this.caption_track) {
                this.language = this.caption_track.language_code;
            }
        }
    }
    /**
     * Deciphers the streaming url of the format.
     * @returns Deciphered URL.
     */
    decipher(player) {
        if (!player)
            throw new InnertubeError('Cannot decipher format, this session appears to have no valid player.');
        return player.decipher(this.url, this.signature_cipher, this.cipher, __classPrivateFieldGet(this, _Format_this_response_nsig_cache, "f"));
    }
}
_Format_this_response_nsig_cache = new WeakMap();
export default Format;
